### ISSUE TITLE
Why

### ISSUE THREAD

**Issue by AlexFundorin** [Erriez/reaper-csurf-juce-template#1]
Hi. What's the purpose of this project?

**Comment by Erriez** [comment]
Hi Alex,

I tried to build a Reaper CSurf surface controller project with Juce GUI's to provide Windows and Mac developers a quick start. The problem is that the Reaper CSurf SDK is out dated (2006), not documented and Windows win32 only. Juce is a great cross-platform development tool, but I could not find any up-to-date CSurf example projects which supports both Windows and Mac. This is requested many times by users on the Reaper forum and this would be a big step forward to support new surfaces in Reaper, supported by the Reaper community as well as the surface manufactures.

In the meantime I've ported the out-dated CSurf SDK (see my other project). I've also reverse engineered all SDK CSurf code and have about 90% code understand of the CSurf Mackie interfaces. However, I'm a Windows developer and I could not find a volunteer to support the Mac Juce part. I'd like to continue with this project when I can share my ideas with someone who's interested to help me/Reaper. Otherwise I'll delete this project and continue with win32 support without Juce and say sorry to Mac users. (Same happened with the extremely out-dated Klinkie CSurf controller)

What do you think about this?

Thanks!

**Comment by AlexFundorin** [comment]
I think that osx users can do what that want to. I'm on Windows myself 😄
I'm trying to understand, what's the point of Jucу GUI in CSurf? Is it meant to be something like Klinke's plugin, with external maps for different controllers/vsts?

**Comment by Erriez** [comment]
Yes, it's to support Mackie or Behringer X-Touch surfaces in Reaper like Klinke's plugin. In this case, Juce is only used to build GUI's with forms, buttons, text boxes, sliders and other custom controls.

Klinke uses (an old fashioned) way of controller mappings. With the current CSurf Reaper API it is possible to create much better mappings and control VST parameters.

Note: The function of a VST plugin is to process MIDI and audio and cannot be used to control surfaces.

**Comment by AlexFundorin** [comment]
In fact, VST **can** be used to control surfaces. For example, Ctrlr app. Ctrlr.org

I'm looking forward for a csurf plugin from you, which would allow to create better mappings and vst control.

**Comment by AlexFundorin** [comment]
There's also VSTLua existing, though it's abandonware and have some major issues when using SYSEX commands in your Lua script. www.hermannseib.com/VstLua/index.html

**Comment by Erriez** [comment]
What kind of surface controller hardware do you have? Do you want to control your VST plugin parameter mappings like Klinke's surface controller in a better way?

My research results:  
Ctrlr is another GUI framework which can be used inside a VST plugin to control MIDI/audio effects/instrument parameters of its own plugin, not to control other VST plugins. For example, Softube Console 1 contains one VST plugin which must be inserted on every track. As a workaround, every Console 1 VST plugin communicates with one application running on the host to control basic track parameters.

A VST2/3 plugin can only retrieve track information and notifications such as volume/pan/name on the current inserted track and not from other tracks. For this reason, Reaper developed a special C++ CSurf framework which also support multiple surface controllers. 

Unfortunately, Lua/VST FX can only use the MIDI port on the inserted track and cannot communicate with other MIDI ports connected to a surface controller. (Confirmed on the Reaper forum some time ago) That's a big limitation, so the Reaper C++ CSurf API is still the only option to control hardware surfaces with a win32, Juce or Ctrl GUI interface.

**Comment by AlexFundorin** [comment]
I own Novation SL MkII. I want to control both VST parameters and Reaper itself.

Ctrlr allows you to control anything via MIDI. It could also control Reaper via OSC, but for now, OSC implementation is broken. I've found out a year ago, while trying to build my own midi<>osc bridge panel for Ctrlr. Here's the link to my topic: http://ctrlr.org/forums/topic/novation-sl-mkii-and-reaper-midiosc-bridge/
The main idea was to control VSTs via OSC in Reaper, but one can also control Reaper via MIDI for now, which means, that VST parameters could be learned. Unfortunately, since reaper doesn't provide midi feedback, this isn't option for me at all. I need feedback to light buttons on my controller and display various info from Reaper on the LCD.

The way Ctrlr works in both instances (standalone and VST) is it opens MIDI ports directly (not through Reaper) and then controls VSTs via OSC

Current implementation of Lua in Reaper doesn't allow full MIDI communication. Only via virtual midid keyboard. I'm aware of this fact. 
So, you're right, there are only two ways to control Reaper for now: via CSurf or via OSC.
There's no software available for now that would let me right a simple math logic script to convert midi<>osc and I'm not an experienced programmer to write such tool myself. I'm hoping that either you ot Geoff Waddington would write such tool in form of CSurf plugin, so it would be possible to achieve my goal.
I've downloaded Visual Studio 2017 today and openframeworks addon, hoping that I could write midi<>osc bridge myself. We'll see, if my attempt will be successful. 
Of cause, I'd prefer a CSurf plugin with customizable maps for midi controllers and preset maps for plugins.

**Comment by Erriez** [comment]
Ok, now I understand what you mean. What you need is a deep Novation SL MkII integration with Reaper by using the CSurf C API. I did some research a while ago to develop this for another device and used scripting as well as OSC. I'll summarize my investigation:

- None of the Reaper scripting languages and VST plugins can be used for surface controller integration. Global MIDI port control is one of the main limitations as well as full control of other track.

- I developed an OSC bridge between Reaper and hardware. With this setup I encountered many OSC limitations which cannot be fixed due to closed source. The Behringer X32 developed a similar bridge, but accepted the limitations.
For example, Kontakt can have >10000 effect parameters. These parameters are divided into a number of FX parameters which are updated over OSC when a FX parameter changes (OSC config DEVICE_FX_PARAM_COUNT and DEVICE_FX_INST_PARAM_COUNT). It is not possible to register on one FX parameter change. Instead, it updates the number of FX parameters (banks) as configured and the device should increment/decrement the FX bank every time. Configuring the number of effect parameters to 10000 results in 10000 OSC messages over UDP when switching between effects. A high performance drop is the consequence, sometimes a Reaper Freeze until all OSC messages are sent. 

In other words: OSC is not designed for surface controllers. The Reaper developers refused to publish OSC as open source.

At the other hand, the Reaper CSurf C API is the only option to control your device completely. With this interface, it is possible to control all MIDI ports, get notifications of every single FX parameter change without performance loss. This allows you to to develop a CSurf plugin for every device to control all buttons/knobs/displays/LED's etc.

Is the complete MIDI implementation of the Novation SL MkII available? That's an important initial step to control every single button/LED/display over MIDI/UDP/TCP/Serial etc. Then I can give you more development hints.

**Comment by AlexFundorin** [comment]
There are .reaperOSC files available and could be edited. I think, that you're aware about them. All OSC commands are there.
Reaper updates every parameter change via OSC and can send parameters of FX or tracks, divided in banks, so it's a bit faster than refreshing the whole surface at once.

Here's MIDI reference for SL MkII. https://www.dropbox.com/s/batcyelpwuj0niq/SLMKII%20MIDI%20Programmers%20Reference.pdf?dl=0

I'm now trying to write oscii-bot script, since Justin updated the app and it's now can operate sysex commands, that are needed for LCD display text.
Here's the link to the forum, where I share my current issues with oscii-bot.
https://forum.cockos.com/showthread.php?p=1913304#post1913304

**Comment by Erriez** [comment]
Wow, you're very lucky with this extended documentation. I had to reverse engineer everything for my device. 

Yes, I fully customized the Default.ReaperOSC file.

> All OSC commands are there.
No, that's not true. Some examples: You cannot fully control the metronome, because several feedback commands are missing in the OSC implementation. It is also not possible to increment/decrement one fader at the time. It requires a complete bank increment/decrement which exists of 8 tracks for a regular console. That was for me a reason to stop the OSC development and switch to CSurf.

> Reaper updates every parameter change via OSC and can send parameters of FX or tracks, divided in banks, so it's a bit faster than refreshing the whole surface at once.
Yes, that's faster, but you cannot display FX parameter EQ low-pass and EQ high-pass when they are located in two different banks. The location of FX parameters inside a bank is unpredictable and requires a bank change every time for one single effect parameter change. If you accept these limitations, OSCII-bot can do the job for you.

**Comment by AlexFundorin** [comment]
> Wow, you're very lucky with this extended documentation. I had to reverse engineer everything for my device.
I was lucky, because, I could convince the support that I needed this document, as well as three others that weren't as useful as this one. Here's the link, if you're interested - https://www.dropbox.com/s/8pkesc4718akqw8/Remote%20SL%20Mk%20II%20Programmers%20reference%20material.zip?dl=0

> You cannot fully control the metronome, because several feedback commands are missing in the OSC implementation.
You mean, like, changing time or tempo? Yeah, that's true.

> It is also not possible to increment/decrement one fader at the time. It requires a complete bank increment/decrement which exists of 8 tracks for a regular console. 
I thought that it's possible to change bank size on fly. Seems like it isn't possible.

> you cannot display FX parameter EQ low-pass and EQ high-pass when they are located in two different banks.
This is why I need some way to store all received data from Reaper and display only those parameters that are needed at the moment.
